File: code_cpp_v2/25.cpp
/*
 * 题目：合并两个排序的链表
 * 链接：https://leetcode-cn.com/problems/he-bing-liang-ge-pai-xu-de-lian-biao-lcof/
 */

#include <iostream>

using namespace std;

struct ListNode {
	int val;
	ListNode *next;

	explicit ListNode(int x) {
		val = x;
		next = nullptr;
	}
};

ListNode *mergeTwoLists(ListNode *l1, ListNode *l2) {
	if (!l1) return l2;
	if (!l2) return l1;

	ListNode* p;
	if (l1->val < l2->val) {
		p = l1;
		p->next = mergeTwoLists(l1->next, l2);
	} else {
		p = l2;
		p->next = mergeTwoLists(l1, l2->next);
	}
	return p;
}

int main() {
	// 手动创建链表
	auto h1 = new ListNode(1);
	auto n11 = new ListNode(2);
	auto n12 = new ListNode(3);
	h1->next = n11;
	n11->next = n12;

	auto h2 = new ListNode(2);
	auto n21 = new ListNode(3);
	auto n22 = new ListNode(4);
	h2->next = n21;
	n21->next = n22;

	// 合并两个链表
	ListNode *newhead = mergeTwoLists(h1, h2);
	ListNode *p = newhead;
	while (p) {
		cout << p->val << " ";
		p = p->next;
	}

	return 0;
}

File: code_swift_v2/25.swift
/*
 * 面试题25：合并两个排序的链表
 * 链接：https://leetcode-cn.com/problems/he-bing-liang-ge-pai-xu-de-lian-biao-lcof/
 */

public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init(_ val: Int) {
        self.val = val
        self.next = nil
    }
}

// MARK: - 解法1
// 递归合并
func mergeTwoLists(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
    if l1 == nil {
        return l2
    }
    if l2 == nil {
        return l1
    }
    
    var p: ListNode? = nil
    if l1!.val < l2!.val {
        p = l1
        p?.next = mergeTwoLists(l1?.next, l2)
    } else {
        p = l2
        p?.next = mergeTwoLists(l1, l2?.next)
    }
    
    return p
}

// main
func main() {
    // 手动创建链表
    let n11: ListNode = ListNode(1)
    let n12: ListNode = ListNode(2)
    let n13: ListNode = ListNode(3)
    let n21: ListNode = ListNode(2)
    let n22: ListNode = ListNode(4)
    let n23: ListNode = ListNode(7)
    n11.next = n12
    n12.next = n13
    n21.next = n22
    n22.next = n23
    
    // 返回链表倒数第k个结点
    let new = mergeTwoLists(n11, n21)
    var p: ListNode? = new
    while p != nil {
        print(p!.val, terminator: "\t")
        p = p?.next
    }
}

main()




